docs/proposals-accepted/202206-active-series-limiting-hashring.md

```diff
@@ -0,0 +1,207 @@
+
+* **Per-tenant limit**: The overall limit imposed for active series per tenant across all replicas of Thanos Receive. This is essentially what this proposal is for.
+
+## Proposal
+
+In general, we would need three measures to impose a limit,
+
+* The current count of active series for a tenant (across all replicas if it is a *per-tenant* limit)
+* The user configured limit (*per-tenant* and *per-replica-tenant* can be different). We can assume this would be available as a user flag and would be same for all tenants (initially)
+* The increase in the number of active series, when a new tenant [remote write request](https://github.com/prometheus/prometheus/blob/v2.36.1/prompb/remote.proto#L22) would be ingested (this can be optional as seen in [meta-monitoring approach](#meta-monitoring-based-validator)).
+
+There are a few ways in which we can achieve the outlined goals of this proposal and get the above measurements to impose a limit. The order of approaches is based on preference.
+
+### Meta-monitoring-based Receive Validator
+
+We could leverage a meta-monitoring solution which scrapes metrics from all Receivers as and then programmatically query for metrics like `prometheus_tsdb_head_series` and sum across tenants and instances periodically and limit based on that value (`latestCurrentSeries`).
+
+This approach would add another mode to Receive apart from Router, Ingestor & RouterIngestor, called **“Validator”**.
+
+Within Validator, we do not need to calculate an increase based on requests, as this will be handled by Receive instrumentation and meta-monitoring solution. We only need to query the latest HEAD series value for a tenant summed across all receives and limit remote write requests if the result of the instant query is greater than the configured limit.
+
+This value can also be cached, and the query for latest value can be executed periodically.
+
+So if a user configures a *per-tenant* limit, say `globalSeriesLimit`, the resultant limiting equation here would simply be `globalSeriesLimit >= latestCurrentSeries` which is checked on request.
+
+<img src="../img/meta-monitoring-validator.png" alt="Meta-monitoring-based Validator" width="800"/>
+
+#### Pros:
+* Simpler as compared to other solutions and easier to implement
+* Lesser endpoint calls, so improved latency
+* Relies on "external to Thanos" system, and doesn’t increase load on Receive
+* Does not add much tenancy-based complexity to Thanos
+* No need to merge statistics across replicas, handled by meta-monitoring
+* Additional request-based rate limiting can be done within same component
+* In case, external meta-monitoring solution is down, can fall back to per-replica-tenant limits
```

Comment:
```suggestion
* In case, external meta-monitoring solution is down, can fall back to per-replica-tenant limits
* Growing our instrumentation to improve validator, improves our observability as well
```